src/generic/FreeAssAlgebraGroebner.jl

```diff
@@ -0,0 +1,640 @@
+    first_term =
+        FreeAssAlgElem{T}(parent(ot.first_poly), ot.first_prefix) *
+        ot.first_poly *
+        FreeAssAlgElem{T}(parent(ot.first_poly), ot.first_suffix)
+    second_term =
+        FreeAssAlgElem{T}(parent(ot.first_poly), ot.second_prefix) *
+        ot.second_poly *
+        FreeAssAlgElem{T}(parent(ot.first_poly), ot.second_suffix)
+    return inv(leading_coefficient(ot.first_poly)) * first_term -
+           inv(leading_coefficient(ot.second_poly)) * second_term
+end
+
+# skip all of the extra length-checking
+function _leading_word(a::FreeAssAlgElem{T}) where {T}
+    return a.exps[1]
+end
+
+@doc """
+    gb_divides_leftmost(a::Word, aut::AhoCorasickAutomaton)
+
+If an element of the Groebner basis that is stored in `aut` divides `a`,
+return (true, a1, a2, keyword_index), where `keyword_index` is the index of the
+keyword that divides `a` such that `a = a1 aut[keyword_index] a2`.
+""" 
+function gb_divides_leftmost(a::Word, aut::AhoCorasickAutomaton)
+    match = search(aut, a)
+    if isnothing(match)
+        return (false, [], [], -1)
+    end
+    return (
+        true,
+        a[1:(match.last_position - length(match.keyword))],
+        a[(match.last_position + 1):end],
+        match.keyword_index,
+    )
+end
+
+# implementation of the normal form function using aho corasick to check for all groebner basis elements in parallel
+@doc """
+    normal_form(f::FreeAssAlgElem{T}, g::Vector{FreeAssAlgElem{T}}, aut::AhoCorasickAutomaton)
+
+Assuming `g` is a groebner basis and `aut` an Aho-Corasick automaton for the elements of `g`,
+compute the normal form of `f` with respect to `g`
+""" 
+function normal_form(
+    f::FreeAssAlgElem{T},
+    g::Vector{FreeAssAlgElem{T}},
+    aut::AhoCorasickAutomaton,
+) where {T}
+    R = parent(f)
+    rexps = Monomial[]
+    rcoeffs = T[]
+    while length(f) > 0
+        ok, left, right, match_index = gb_divides_leftmost(f.exps[1], aut)
+        if ok
+            qi = divexact(f.coeffs[1], g[match_index].coeffs[1])
+            f = _sub_rest(f, mul_term(qi, left, g[match_index], right), 1)
+        else
+            push!(rcoeffs, f.coeffs[1])
+            push!(rexps, f.exps[1])
+            f = FreeAssAlgElem{T}(R, f.coeffs[2:end], f.exps[2:end], length(f) - 1)
+        end
+    end
+    return FreeAssAlgElem{T}(R, rcoeffs, rexps, length(rcoeffs))
+end
+
+# normal form with leftmost word divisions
+function normal_form(
+    f::FreeAssAlgElem{T},
+    g::Vector{FreeAssAlgElem{T}},
+) where {T<:FieldElement}
+    R = parent(f)
+    s = length(g)
+    rcoeffs = T[]
+    rexps = Monomial[]
+    while length(f) > 0
+        i = 1
+        @label again
+        ok, ml, mr = word_divides_leftmost(f.exps[1], g[i].exps[1])
+        if !ok && i < s
+            i += 1
+            @goto again
+        end
+        if ok
+            qi = divexact(f.coeffs[1], g[i].coeffs[1])
+            f = _sub_rest(f, mul_term(qi, ml, g[i], mr), 1) # enforce lt cancelation
```

Comment:
This implementation of `normal_form` can be made much more efficient. You are turning a part that could be (almost) linear into a quadratic one, and generating tons of useless allocations in the process.

I think I pointed out at some point in the past at the article http://arxiv.org/abs/1206.6940v1 which I still think is a fantastic reference -- there, take a look at section 4, specifically 4.1, to find an easy algorithm for making this quite a bit faster (using a heap and a hash table -- Geobuckets are more fancy but probably overkill here)

Reply:
What part exactly do you mean is currently quadratic instead of almost linear? Is the problem that right now the terms of an element are always completely ordered when doing a multiplication?

Reply:
I started looking into this but didn't get to finish it yet and now I will be on vacation for a few weeks.
Would you say this optimization is crucial for the current pull request or would it be alright if I got back to this after my vacation and maybe open a new pull request when it is finished?

Reply:
I don't think it is crucial